### PR TITLE
[CM T1] Improve and extend the tutorial for writing specifications

### PR DESCRIPTION
**Pull request by flo91**
This T1-PR was done by me, Florian Lindner.

## Basics

- [x] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [x] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [x] References to issues, e.g. `close #X`, are in the commit messages.
- [x] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [x] The PR is rebased with current master.

## Checklist

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned [every code](/.reuse/dep5) not directly written by me in [reuse syntax](https://reuse.software/)

## Review

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

- [ ] Add the "work in progress" label if you do not want the PR to be reviewed yet.
- [ ] Add the "ready to merge" label **if the basics are fulfilled** and no further pushes are planned by you.

### CODE REVIEW

**Review by kodebach — CHANGES_REQUESTED** [review]
Great work :+1: 

I have some suggestions to improve readability. There is also some overlap with #4319.

**Comment by kodebach** [comment]
After talking to @markus2330, please also make sure that there is a shell recorder test for the specification tutorial (like `testshell_markdown_tutorial_merge` for the merge tutorial). It should just be a case of adding the CMake call to `add_msr_test` in the correct place...

**Comment by markus2330** [comment]
... and the correct place is tests/shell/shell_recorder/tutorial_wrapper/CMakeLists.txt :rocket:

**Comment by kodebach** [comment]
Regarding the failed `alpine` job: It seems the `date` plugin is explicitly excluded on Alpine Linux

https://github.com/ElektraInitiative/libelektra/blob/f2b9d555a2de6da944781bd9edb4a2a877cd3c35/scripts/jenkins/Jenkinsfile#L691

I don't know why, I could try to include it. If it really doesn't work, you can either change the tutorial to use a different plugin, or exclude the test on Alpine (via CMake).

As for the ASAN failure: I think we need to add a new `no_sanitize` attribute to the hash function. Specifically, I think we need to add `no_sanitize("implicit-unsigned-integer-truncation")`. To do this, add a new CMake variable here:

https://github.com/ElektraInitiative/libelektra/blob/5702ce0496823fdd12c4e1e6ff9d6d1ca1b71a40/src/include/CMakeLists.txt#L138-L155

Then add the new `#define` here:

https://github.com/ElektraInitiative/libelektra/blob/5702ce0496823fdd12c4e1e6ff9d6d1ca1b71a40/src/include/kdbconfig.h.in#L61-L65

And finally, use the new attribute here:

https://github.com/ElektraInitiative/libelektra/blob/f2b9d555a2de6da944781bd9edb4a2a877cd3c35/src/plugins/ni/nickel-1.1.0/src/hash.c#L329-L331

and here:

https://github.com/ElektraInitiative/libelektra/blob/f2b9d555a2de6da944781bd9edb4a2a877cd3c35/src/plugins/ni/nickel-1.1.0/src/hash.c#L888-L890

File: tests/shell/shell_recorder/tutorial_wrapper/CMakeLists.txt [hunk 1]
```diff
@@ -32,6 +32,7 @@ add_msr_test (
 	ENVIRONMENT KDB_EXEC_PATH=${CMAKE_SOURCE_DIR}/scripts/kdb)
 add_msr_test (tutorial_storage_plugins "${CMAKE_SOURCE_DIR}/doc/tutorials/storage-plugins.md" REQUIRED_PLUGINS type yamlcpp)
 add_msr_test (cmerge "${CMAKE_SOURCE_DIR}/doc/tutorials/cmerge.md" REQUIRED_PLUGINS hosts line PROPERTY LABELS memleak)
+add_msr_test (tutorial_spec "${CMAKE_SOURCE_DIR}/doc/tutorials/specification.md")
```
Comment (kodebach) [inline tests/shell/shell_recorder/tutorial_wrapper/CMakeLists.txt:35]:
```suggestion
add_msr_test (tutorial_spec "${CMAKE_SOURCE_DIR}/doc/tutorials/specification.md" PROPERTY LABELS memleak)
```

Seems like adding the `NO_SANITIZE` attributes didn't help. Please exclude the test from ASAN.

**Comment by flo91** [comment]
jenkins build libelektra please

**Pull request merged**